modyn/supervisor/internal/eval/handler.py

```diff
@@ -83,56 +85,88 @@ def get_eval_requests_after_pipeline(self, df_trainings: pd.DataFrame) -> list[E
+            columns=["start", "end", "active_model_trained_before"],
         )
 
         # now build & potentially reduce the Trainings x EvalIntervals space
         df_cross = df_eval_intervals.merge(df_trainings, how="cross")
 
-        # to check if a combination is the most recent, we first compute if model was trained before the usage starts
-        # last_timestamp (df_trainings) defines the end of the training data;
-        # most_recent_model_interval_end_before (df_eval_intervals): defines center of an eval intervals.
-        df_cross["chronology_ok"] = df_cross["last_timestamp"] <= df_cross["most_recent_model_interval_end_before"]
+        # to check if a combination is the most recent or currently, we first compute if model was trained before
+        # the usage starts last_timestamp (df_trainings) defines the end of the training data;
+        # active_model_trained_before (df_eval_intervals): defines center of an eval intervals.
+        df_cross["active_candidate"] = df_cross["real_last_timestamp"] < df_cross["active_model_trained_before"]
 
         # find the maximum model for every EvalCandidate that doesn't violate that constraint
         max_model_id = (
-            df_cross[df_cross["chronology_ok"]]
-            .groupby("most_recent_model_interval_end_before")["id_model"]
+            df_cross[df_cross["active_candidate"]]
+            .groupby("active_model_trained_before")["id_model"]
             .aggregate(max_model_id="max")
         )
 
         # combine: a model in the cross product is most recent for a certain interval iff
-        #  it has maximum model id for its most_recent_model_interval_end_before
-        final = df_cross.merge(max_model_id, on="most_recent_model_interval_end_before")
-        final["most_recent_model"] = final["id_model"] == final["max_model_id"]
+        #  it has maximum model id for its active_model_trained_before
+        df_active_models = df_cross.merge(max_model_id, on="active_model_trained_before", how="left")
+        df_active_models["active_model"] = df_active_models["id_model"] == df_active_models["max_model_id"]
+
+        # for a given interval, the currently trained model is the model with the smallest id
+        # from all models that have a strictly bigger id than the most recent model. Hence it is the model after the
+        # most recent model.
+        # For that we first build a model -> successor model mapping:
+        model_successor_relation = df_active_models[["id_model"]].drop_duplicates().sort_values(by="id_model")
+        model_successor_relation["next_id_model"] = model_successor_relation["id_model"].shift(-1, fill_value=-1)
+
+        # if there's no active model for the first interval(s), we still need to define the next model as the
+        # trained model
+        model_successor_relation = pd.concat(
+            [model_successor_relation, pd.DataFrame([{"id_model": None, "next_id_model": 21}])]
```

Comment:
Where is the 21 coming from?

Reply:
Uff, that shouldn't be there, artifact debugging... I wanted to use the minimum model as it has no predecessor.

Reply:
Ok perfect!